### DISCUSSION TITLE
cuTensorNet is slower than opt_einsum for QFT and QAOA circuits

### DISCUSSION THREAD

**Discussion by rht** [NVIDIA/cuQuantum#53] [Q&A] [answered]
I was benchmarking cuTensorNet and opt_einsum on the QFT and QAOA circuit, and I found the former to be consistently slower than the latter.
I used the same code in https://github.com/NVIDIA/cuQuantum/discussions/23 for cuTensorNet, and for opt_einsum:
```python
bitstring = "0" * len(qubits)
# https://optimized-einsum.readthedocs.io/en/stable/autosummary/opt_einsum.contract_path.html#opt_einsum.contract_path
expression, operands = myconverter.amplitude(bitstring=bitstring)
tic = time.time()
path, path_info = oe.contract_path(expression, *operands)
elapsed1 = time.time() - tic
print("Elapsed opt_einsum path finding", elapsed1)
tic = time.time()
output = oe.contract(expression, *operands, optimize=path)
elapsed2 = time.time() - tic
print("Elapsed opt_einsum contract", elapsed2)
```

Plot for QFT:
![benchmark](https://github.com/NVIDIA/cuQuantum/assets/395821/ee09fea1-4d13-46d6-a081-770bbe516c55)

I think it has something to do with cuTensorNet's path finding being closer to the global optimum. Is there a way to tweak the hyperoptimizer to either stop early, or to have a larger error tolerance, so as to minimize the overall time?

**Comment by rht** [comment]
This is QAOA for path only. Can't find the plot where I did both `contract_path` and `contract`, but I remembered `cutn` still being slower for QAOA.
![benchmark_qaoa](https://github.com/NVIDIA/cuQuantum/assets/395821/af8d4d6e-e168-4c8c-981b-7dc625e259b5)

**Reply by rht** [reply]
The cuQuantum Appliance version I have used is 23.03.

**Comment by leofang** [comment]
Hi @rht, thanks for reporting and sorry for not responding promptly. We were impacted by a global notification setting change across all NV repos.

First, could you kindly offer the full benchmark script for cutn? The discussion in #23 was fragmented, and it's easier for us to give advices after reading the full script.

**Reply by rht** [reply]
OK, here is the code: https://gist.github.com/rht/b65dc22ff4826170710661a4e5fec29a. Apologize in advance if it is spaghetti code.
L51-102 is the relevant `cutn` code. I had tried varying the configuration I can find from the API, with no observable result in reducing the path finding time. The `opt_einsum` code is at L166-181

Summary of the recommendation from #23:
- I made sure to create a reusable handle in the beginning
- I used `blocking="auto"`
- I separated the `contract` step into `contract_path` and `contract`.

**Reply by rht** [reply]
I have simplified the benchmark code. I removed Cotengra, and I restricted to only calculating the expectation value of Pauli Z on the first qubit. It should be more readable now.

**Comment by rht** [comment]
This is related to https://github.com/NVIDIA/cuQuantum/discussions/53, but instead, the reason why `cutn` was slower in this benchmark, was that the path finding time was longer. It might have been fixed in 23.06, but I'm waiting for the cuQuantum Appliance to be available.

**Reply by rht** [reply]
This is the new benchmark with cuQuantum 23.06. `cutn` is now much faster, but still slightly slower than `opt_einsum`.

![vqe_qaoa_linear](https://github.com/NVIDIA/cuQuantum/assets/395821/71482f84-c061-4595-a69c-d9401ac91d5d)

**Reply by rht** [reply]
The circuit is `ansatz = TwoLocal(num_qubits, "ry", "cz", reps=5, entanglement=entanglement)`. I use the default `contract` from `cutn` without splitting into `contract_path` and `contract`. Likewise for `opt_einsum`.

**Comment by mtjrider** [comment]
Hi @rht,

Is this performance regression still relevant?
Would you mind sharing an update so that we can prioritize accordingly?

Thanks in advance.

**Reply by rht** [reply]
It is faster on 23.06, but still slower than opt_einsum. See https://github.com/NVIDIA/cuQuantum/discussions/53#discussioncomment-6474186.

**Reply by mtjrider** [reply]
Thank you for the update.
I will be locking this discussion, and creating an issue based on the feedback you've provided.

**Comment by mtjrider** [comment, marked as answer]
I've created an issue to track the apparent performance issue.

Link here: https://github.com/NVIDIA/cuQuantum/issues/90
